controllers/controllers.go

```diff
@@ -41,19 +41,88 @@ type instanceReconciler struct {
 	recorder record.EventRecorder
 }
 
-// configureInstance adds the specified instance to the cluster. if hostname is not empty, the instance's hostname will be
-// changed to the passed in value. If annotations is not nil, the node will have the specified annotations applied to
-// it.
-func (r *instanceReconciler) configureInstance(instance *instances.InstanceInfo, annotations map[string]string) error {
+// ensureInstanceIsUpToDate ensures that the given instance is configured as a node and upgraded to the specifications
+// defined by the current version of WMCO. If annotations is not nil, the node will have the specified annotations
+// applied to it.
+func (r *instanceReconciler) ensureInstanceIsUpToDate(instance *instances.InstanceInfo, annotationsToApply map[string]string) error {
+	if instance == nil {
+		return errors.New("instance cannot be nil")
+	}
+
+	// Instance is up to date, do nothing
+	upToDate, err := instanceUpToDate(instance)
+	if err != nil {
+		return err
+	}
+	if upToDate {
+		return nil
+	}
+
 	nc, err := nodeconfig.NewNodeConfig(r.k8sclientset, r.clusterServiceCIDR, r.vxlanPort, instance, r.signer,
-		annotations)
+		annotationsToApply)
 	if err != nil {
 		return errors.Wrap(err, "failed to create new nodeconfig")
 	}
-	if err := nc.Configure(); err != nil {
-		return errors.Wrap(err, "failed to configure Windows instance")
+
+	// Check if the instance was configured by a previous version of WMCO and must be deconfigured before being
+	// configured again.
+	upgradeRequired, err := instanceUpgradeRequired(instance)
+	if err != nil {
+		return err
+	}
+	if upgradeRequired {
+		if err := nc.Deconfigure(); err != nil {
+			return err
+		}
 	}
-	return nil
+
+	return nc.Configure()
+}
+
+// instanceUpToDate returns true if the instance was configured by the current WMCO version
+func instanceUpToDate(instance *instances.InstanceInfo) (bool, error) {
+	if instance == nil {
+		return false, errors.New("instance cannot be nil")
+	}
+
+	if instance.Node == nil {
+		return false, nil
+	}
+
+	versionAnnotation, present := instance.Node.GetAnnotations()[nodeconfig.VersionAnnotation]
+	return present && versionAnnotation == version.Get(), nil
+}
+
+// instanceUpgradeRequired returns true if the instance needs to go through the upgrade process.
+func instanceUpgradeRequired(instance *instances.InstanceInfo) (bool, error) {
```

Comment:
Sorry @sebsoto , I was not thinking things through with my previous [comment](https://github.com/openshift/windows-machine-config-operator/pull/513#discussion_r673283175). These functions should actually be part of the `instance` struct. But given it was an oversight on my part, I can open a follow up PR if you just want to get this merged.

Reply:
If you decide to make the change please ensure the calls don't stutter.

Reply:
In order to do that the VersionAnnotation const would need to pulled out of the nodeconfig package, as currently trying to use it from the `instances` package would result in a circular dependency. I can give it a shot.
I don't want to do the alternative which would be passing in the annotation key to the NewInstance function.